### ISSUE TITLE
Different results after DDIM inverse.

### ISSUE THREAD

**Issue by Sumching** [huggingface/diffusers#4654]
`
        from diffusers import DDIMScheduler, UNet2DModel, DDIMInverseScheduler
        import torch
        import numpy as np
        import cv2
        from tqdm import tqdm
        
        def ddpm_forward(noise, model, scheduler, timesteps=500):
            scheduler.set_timesteps(timesteps)
            sample_size = model.config.sample_size
            input = torch.tensor(noise).to("cuda")
        
            for t in tqdm(scheduler.timesteps):
                with torch.no_grad():
                    noisy_residual = model(input, t).sample
                previous_noisy_sample = scheduler.step(noisy_residual, t, input).prev_sample
                input = previous_noisy_sample
        
        
            return input
        
        
        def ddpm_backward(x0, model, scheduler, timesteps=500):
            scheduler.set_timesteps(timesteps)
            sample_size = model.config.sample_size
            noise = torch.tensor(x0).to("cuda")
            input = noise
        
            for t in tqdm(scheduler.timesteps):
                with torch.no_grad():
                    noisy_residual = model(input, t).sample
                previous_noisy_sample = scheduler.step(noisy_residual, t, input).prev_sample
                input = previous_noisy_sample
        
            
        
            # image = (input / 2 + 0.5).clamp(0, 1).squeeze()
            # image = (image.permute(1, 2, 0) * 255).round().to(torch.uint8).cpu().numpy()
            return input
        
        
        
        scheduler = DDIMScheduler.from_pretrained("google/ddpm-cifar10-32", cache_dir="/data/diffuser_model")
        invscheduler = DDIMInverseScheduler.from_pretrained("google/ddpm-cifar10-32", cache_dir="/data/diffuser_model")
        model = UNet2DModel.from_pretrained("google/ddpm-cifar10-32", use_safetensors=True, cache_dir="/data/diffuser_model").to("cuda")
        
        
        noise = np.random.randn(1, 3, 32, 32).astype("float32")
        x0 = ddpm_forward(noise, model, scheduler)
        oimg = (x0 / 2 + 0.5).clamp(0, 1).squeeze()
        oimg = (oimg.permute(1, 2, 0) * 255).round().to(torch.uint8).cpu().numpy()
        cv2.imwrite("ddim_gen.jpg", oimg)
         
        
        
        latent = ddpm_backward(x0, model, invscheduler)
        
        nimg = ddpm_forward(latent, model, scheduler)
        nimg = (nimg / 2 + 0.5).clamp(0, 1).squeeze()
        nimg = (nimg.permute(1, 2, 0) * 255).round().to(torch.uint8).cpu().numpy()
        cv2.imwrite("ddim_re.jpg", nimg)
        
        err = oimg/255 - nimg/255
        import pdb;pdb.set_trace()

`

the nimg looks very different.

**Comment by Devanshi-Crypto** [comment]
Hey, can you explain this issue to me? I'd like to contribute

**Comment by Sumching** [comment]
First, I randomly generate an image 'oimg' through a forward process. Then, I inverse the 'oimg' to obtain its latent code. Finally, by feeding the latent code into the forward function, I aim to obtain the reconstructed image 'nimg' such that nimg is equal to oimg.

**Comment by sayakpaul** [comment]
Cc: @clarencechen and @patrickvonplaten

**Comment by tvaranka** [comment]
Inversion using DDIM is expected to lead to insufficient reconstruction of the original input as it relies on local linearization [1]. 

For exact solution, EDICT is able to provide accurate reconstruction. It is worth noting that the there is still error due to the compression from VQGAN. EDICT is however quite slow and requires two latent codes. New exact and lossy inversion techniques are popping up quite frequently however, so keep an eye out for them.

Figure from EDICT paper

 
![image](https://github.com/huggingface/diffusers/assets/32382096/14eeadd3-4fd9-40a2-baa4-af2ca1a3fe0c)

[1] Wallace, Bram, Akash Gokul, and Nikhil Naik. "Edict: Exact diffusion inversion via coupled transformations." Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition. 2023.
